### PR TITLE
CLOUD-424

### PR DESCRIPTION
**Pull request by nonemax**

### CODE REVIEW

File: integtests/apitester/apitester.go [hunk 1]
```diff
@@ -50,57 +43,101 @@ func (c *Config) Run() error {
...
+	if err != nil {
+		return err
 	}
+
 	return nil
 }
 
-func (c *Config) checkStatus(ts *httptest.Server, testCase Case) error {
+type GetInstanceResp struct {
+	LastOperatiom struct {
```
Comment (dAdAbird) [inline integtests/apitester/apitester.go]: s/LastOperatiom/LastOperatio*n*

File: integtests/apitester/apitester.go [hunk 2]
```diff
@@ -50,57 +43,101 @@ func (c *Config) Run() error {
...
+			fmt.Println()
+			ticker.Stop()
+			if status != testCase.RespStatus {
+				return errors.New("Wrong resp status")
+			}
+			err := checkRespData(testCase.RespData, respData)
+			if err != nil {
+				return err
+			}
 			return nil
```
Comment (dAdAbird) [inline integtests/apitester/apitester.go]: just `return err`

File: integtests/apitester/apitester.go [hunk 3]
```diff
@@ -50,57 +43,101 @@ func (c *Config) Run() error {
...
-		respData, _, err := c.Request(ts.URL, testCase.ReqType, testCase.ReqData)
+		respData, status, err := c.Request(c.Address+testCase.Endpoint, testCase.ReqType, testCase.ReqData)
 		if err != nil {
 			return err
 		}
 		err = json.Unmarshal(respData, &instResp)
 		if err != nil {
-			return err
+			log.Println(err)
+			continue
```
Comment (dAdAbird) [inline integtests/apitester/apitester.go:83]: why we log error here instead of return?

File: integtests/apitester/apitester.go [hunk 4]
```diff
@@ -50,57 +43,101 @@ func (c *Config) Run() error {
...
 		}
-		if instResp.State != "succeeded" {
+
+		if instResp.LastOperatiom.State == "succeeded" {
+			fmt.Println()
+			ticker.Stop()
+			if status != testCase.RespStatus {
+				return errors.New("Wrong resp status")
```
Comment (dAdAbird) [inline integtests/apitester/apitester.go]: it's better to have more info on which exactly status should be and what was got instead

File: integtests/apitester/apitester.go [hunk 5]
```diff
@@ -50,57 +43,101 @@ func (c *Config) Run() error {
...
+			fmt.Println()
 			return fmt.Errorf("cluster not started")
 		}
 	}
-	ticker.Stop()
 
 	return nil
 }
 
+func checkRespData(waitData, respData []byte) error {
```
Comment (dAdAbird) [inline integtests/apitester/apitester.go]: s/waitData/expected

File: integtests/apitester/apitester.go [hunk 6]
```diff
@@ -50,57 +43,101 @@ func (c *Config) Run() error {
...
 		if err != nil {
-			return err
+			log.Println(err)
+			continue
 		}
-		if instResp.State != "succeeded" {
+
+		if instResp.LastOperatiom.State == "succeeded" {
+			fmt.Println()
+			ticker.Stop()
```
Comment (dAdAbird) [inline integtests/apitester/apitester.go]: it's more secure to put ticker.Stop() into defer right after the ticker initialising

File: integtests/apitester/apitester.go [hunk 7]
```diff
@@ -50,57 +43,101 @@ func (c *Config) Run() error {
...
+	LastOperatiom struct {
+		State       string `json:"state"`
+		Description string `json:"description"`
+	} `json:"last_operation"`
+}
+
+func (c *Config) checkStatus(testCase structs.CaseData) error {
 	startTime := time.Now()
-	endTime := startTime.Local().Add(time.Second * time.Duration(180))
+	endTime := startTime.Local().Add(time.Second * time.Duration(250))
```
Comment (dAdAbird) [inline integtests/apitester/apitester.go]: 1. `time.Now()` already returns `local time`. So no need in `startTime.Local()`. Anyway, it affect neither wall nor monotonic time so `time.Now().UTC().Unix()` always `time.Now().Local().Unix()`. It just makes a confusion here.
2. time.Second * 250

File: integtests/apitester/apitester.go [hunk 8]
```diff
@@ -50,57 +43,101 @@ func (c *Config) Run() error {
...
 		err = json.Unmarshal(respData, &instResp)
 		if err != nil {
-			return err
+			log.Println(err)
+			continue
 		}
-		if instResp.State != "succeeded" {
+
+		if instResp.LastOperatiom.State == "succeeded" {
+			fmt.Println()
```
Comment (dAdAbird) [inline integtests/apitester/apitester.go]: ?

File: integtests/apitester/apitester.go [hunk 9]
```diff
@@ -3,28 +3,21 @@ package apitester
...
 	"fmt"
 	"io/ioutil"
 	"log"
 	"net/http"
-	"net/http/httptest"
 	"time"
 
-	"github.com/Percona-Lab/percona-dbaas-cli/broker"
 	"github.com/Percona-Lab/percona-dbaas-cli/integtests/datafactory"
+	"github.com/Percona-Lab/percona-dbaas-cli/integtests/structs"
```
Comment (dAdAbird) [inline integtests/apitester/apitester.go:14]: I don't see this package in the repo

File: integtests/datafactory/datafactory.go [hunk 10]
```diff
@@ -22,46 +17,69 @@ func GetCreatePXCInstanceData() CaseData {
...
-		ReqType:  "GET",
-		ReqData:  []byte(``),
-		RespData: ``,
+func GetGetPXCInstanceData() structs.CaseData {
+	return structs.CaseData{
+		Endpoint:   "/v2/service_instances/test-pxc-instance",
+		ReqType:    "GET",
+		ReqData:    []byte(``),
+		RespStatus: 200,
+		RespData:   []byte(`{"id":"test-pxc-instance","dashboard_url":"","internalId":"test-pxc-instance","service_id":"percona-xtradb-cluster-id","plan_id":"percona-xtradb-id","organization_guid":"","space_guid":"","last_operation":{"state":"succeeded","description":"successfully created service instance","async_poll_interval_seconds":10},"parameters":{"cluster_name":"test-pxc","replicas":3,"proxy_sql_replicas":1,"topology_key":"none","size":"1Gi"},"context":{},"credentials":{"host":"","port":0,"users":{"clustercheck":"dkpPNjNYTUZUYmJyaldRRkNz","monitor":"MDU3QWFjcDJZOXdTMjg5M0pz","proxyadmin":"UlFQWUg1dTNjdWpSU0FaUzFKMQ==","root":"djBSOENpeUk3TWU0T1pBOVlR","xtrabackup":"anlRMVZiUkdyT09GRXJIdHRDWA=="}},"credentialData":{"message":"MySQL cluster started successfully","host":"test-pxc-proxysql","port":3306,"user":"root","pass":"v0R8CiyI7Me4OZA9YQ"}}`),
```
Comment (dAdAbird) [inline integtests/datafactory/datafactory.go]: Why keep this data as a []byte if using it always assumes marshalling into `structs.ServiceInstance`?

File: integtests/apitester/apitester.go [hunk 11]
```diff
@@ -117,83 +154,22 @@ func (c *Config) Request(address, reqType string, reqBody []byte) ([]byte, int,
...
 
-type GetInstanceResp struct {
-	State       string `json:"state"`
-	Description string `json:"description"`
+	t.Cases = append(t.Cases, datafactory.GetCreatePSMDBInstanceData())
+	t.Cases = append(t.Cases, datafactory.GetGetPSMDBInstanceData())
+	t.Cases = append(t.Cases, datafactory.GetUpdatePSMDBInstanceData())
+	t.Cases = append(t.Cases, datafactory.GetGetPSMDBInstanceUpdatedData())
+	t.Cases = append(t.Cases, datafactory.GetDeletePSMDBInstanceData())
+	t.Cases = append(t.Cases, datafactory.GetGetDeletedPSMDBInstanceData())
```
Comment (dAdAbird) [inline integtests/apitester/apitester.go]: What a need in this datafactory? why not keep cases as a slice?
Why add extra logic - new object, set, etc.? When it's enough to have a global slice of cases and iterate thru it in function (not a method)  "Run(serverAddress)"

**Review by dAdAbird — CHANGES_REQUESTED** [review]

File: integtests/apitester/apitester.go [hunk 12]
```diff
@@ -50,57 +43,88 @@ func (c *Config) Run() error {
...
 		if err != nil {
-			return err
+			log.Println(err)
+			continue
 		}
-		if instResp.State != "succeeded" {
-			return nil
+
+		if instResp.LastOperation.State == "succeeded" {
+			fmt.Println() // just for create new line after ticker
```
Comment (dAdAbird) [inline integtests/apitester/apitester.go]:
```suggestion
			fmt.Println() // print a new line after the ticker
```

File: integtests/apitester/apitester.go [hunk 13]
```diff
@@ -50,57 +43,88 @@ func (c *Config) Run() error {
...
+			if status != testCase.RespStatus {
+				return errors.New("Wrong resp status. Have '" + instResp.LastOperation.State + "', want 'succeeded'")
+			}
+			return checkRespData(testCase.RespData, respData)
 		}
+		fmt.Printf("\r Wait for cluster. %v tries left  ", (endTime.Unix()-t.Unix())/2)
 		if t.Unix() >= endTime.Unix() {
-			ticker.Stop()
+			fmt.Println() // just for create new line after ticker
```
Comment (dAdAbird) [inline integtests/apitester/apitester.go]:
```suggestion
			fmt.Println() // print a new line after the ticker
```

File: integtests/apitester/apitester.go [hunk 14]
```diff
@@ -50,57 +43,88 @@ func (c *Config) Run() error {
...
-func (c *Config) checkStatus(ts *httptest.Server, testCase Case) error {
+type GetInstanceResp struct {
+	LastOperation struct {
+		State       string `json:"state"`
+		Description string `json:"description"`
+	} `json:"last_operation"`
+}
+
+func (c *Config) waitForSucceed(testCase structs.CaseData) error {
 	startTime := time.Now()
```
Comment (dAdAbird) [inline integtests/apitester/apitester.go]:
```suggestion

```

File: integtests/apitester/apitester.go [hunk 15]
```diff
@@ -50,57 +43,88 @@ func (c *Config) Run() error {
...
-	endTime := startTime.Local().Add(time.Second * time.Duration(180))
+	endTime := startTime.Add(time.Second * 250)
```
Comment (dAdAbird) [inline integtests/apitester/apitester.go]:
```suggestion
	endTime := time.Now().Add(time.Second * 250)
```

File: integtests/structs/instance.go [hunk 16]
```diff
@@ -0,0 +1,72 @@
+package structs
+
+const (
+	InProgressOperationSate        OperationState       = "in progress"
+	InProgressOperationDescription OperationDescription = "creating service instance..."
+	SucceedOperationState          OperationState       = "succeeded"
+	SucceedOperationDescription    OperationDescription = "successfully created service instance"
+	FailedOperationState           OperationState       = "failed"
+	FailedOperationDescription     OperationDescription = "failed to create service instance"
+)
```
Comment (dAdAbird) [inline integtests/structs/instance.go:8]: 1. A common style is declaring custom type right above constants that use it. And to group constants by type. Besides no need to declare type along with each const.
```go
type OperationState string

const (
	InProgressOperationSate OperationState = "in progress"
	SucceedOperationState                  = "succeeded"
	FailedOperationState                   = "failed"
)
...
```

2. I don't see any usage of those constants in code. Use them!)

File: integtests/apitester/apitester.go [hunk 17]
```diff
@@ -50,57 +43,88 @@ func (c *Config) Run() error {
...
+	}
+	if respInst.Parameters.Replicas != expected.Parameters.Replicas {
+		return errors.New("wrong replicas number")
+	}
+	if respInst.Parameters.ClusterName != expected.Parameters.ClusterName {
+		return errors.New("wrong cluster name")
+	}
+	if respInst.ID != expected.ID {
+		return errors.New("wrong ID")
+	}
```
Comment (dAdAbird) [inline integtests/apitester/apitester.go]: 1. Need to pad on what exactly went wrong (what we got and what was expected)
2. Expected data in different cases have a different number of fields. But you make the same checks for each case. If this is the only data that should be checked so no need in extra fields inside the data factory. Or we need a custom check?

File: integtests/apitester/apitester.go [hunk 18]
```diff
@@ -50,57 +43,88 @@ func (c *Config) Run() error {
...
+			if status != testCase.RespStatus {
+				return errors.New("Wrong resp status. Have '" + instResp.LastOperation.State + "', want 'succeeded'")
+			}
+			return checkRespData(testCase.RespData, respData)
 		}
+		fmt.Printf("\r Wait for cluster. %v tries left  ", (endTime.Unix()-t.Unix())/2)
 		if t.Unix() >= endTime.Unix() {
-			ticker.Stop()
+			fmt.Println() // just for create new line after ticker
 			return fmt.Errorf("cluster not started")
```
Comment (dAdAbird) [inline integtests/apitester/apitester.go]:
```suggestion
			return errors.New("cluster not started")
```

File: integtests/apitester/apitester.go [hunk 19]
```diff
@@ -50,57 +43,88 @@ func (c *Config) Run() error {
...
-			return nil
+
+		if instResp.LastOperation.State == "succeeded" {
+			fmt.Println() // just for create new line after ticker
+			if status != testCase.RespStatus {
+				return errors.New("Wrong resp status. Have '" + instResp.LastOperation.State + "', want 'succeeded'")
+			}
+			return checkRespData(testCase.RespData, respData)
 		}
+		fmt.Printf("\r Wait for cluster. %v tries left  ", (endTime.Unix()-t.Unix())/2)
```
Comment (dAdAbird) [inline integtests/apitester/apitester.go:93]: Are you sure there is should be `\r = CR (Carriage Return)`?

File: integtests/apitester/apitester.go [hunk 20]
```diff
@@ -50,57 +43,88 @@ func (c *Config) Run() error {
...
+			log.Println(err)
+			continue
 		}
-		if instResp.State != "succeeded" {
-			return nil
+
+		if instResp.LastOperation.State == "succeeded" {
+			fmt.Println() // just for create new line after ticker
+			if status != testCase.RespStatus {
+				return errors.New("Wrong resp status. Have '" + instResp.LastOperation.State + "', want 'succeeded'")
```
Comment (dAdAbird) [inline integtests/apitester/apitester.go]: 1. "Error strings should not be capitalized"
2. It is definitely something wrong here. You check `status` and `testCase.RespStatus`, but the error contain quite a bit different statements.

File: integtests/apitester/apitester.go [hunk 21]
```diff
@@ -50,57 +43,88 @@ func (c *Config) Run() error {
...
+	endTime := startTime.Add(time.Second * 250)
 	ticker := time.NewTicker(2 * time.Second)
-
+	defer ticker.Stop()
 	var instResp GetInstanceResp
 	for t := range ticker.C {
-		respData, _, err := c.Request(ts.URL, testCase.ReqType, testCase.ReqData)
+		respData, status, err := c.Request(c.Address+testCase.Endpoint, testCase.ReqType, testCase.ReqData)
 		if err != nil {
 			return err
```
Comment (dAdAbird) [inline integtests/apitester/apitester.go]: Add some context to the error.

File: integtests/apitester/apitester.go [hunk 22]
```diff
@@ -50,57 +43,88 @@ func (c *Config) Run() error {
...
 	}
-	respData, respStatus, err := c.Request(ts.URL, testCase.ReqType, testCase.ReqData)
+	respData, respStatus, err := c.Request(c.Address+testCase.Endpoint, testCase.ReqType, testCase.ReqData)
 	if err != nil {
 		return err
 	}
-	if respStatus != http.StatusOK || respStatus != http.StatusAccepted {
-		log.Println(respStatus)
+	if respStatus != testCase.RespStatus {
 		return fmt.Errorf("Wrong status")
```
Comment (dAdAbird) [inline integtests/apitester/apitester.go]:
```suggestion
		return fmt.Errorf("wrong status")
```

**Review by dAdAbird — CHANGES_REQUESTED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| psmdb | failed |
| pxc | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| pxc | failed |
| psmdb | failed |
| service-broker | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| pxc | failed |
| service-broker | passed |
| psmdb | failed |

File: integtests/apitester/apitester.go [hunk 23]
```diff
@@ -42,158 +36,127 @@ func (c *Config) Run() error {
...
 		err = json.Unmarshal(respData, &instResp)
 		if err != nil {
-			return err
+			log.Println(err)
+			continue
 		}
-		if instResp.State != "succeeded" {
-			return nil
+
+		if instResp.LastOperation.State == "succeeded" {
```
Comment (dAdAbird) [inline integtests/apitester/apitester.go]:
```suggestion
		if instResp.LastOperation.State == structs.SucceedOperationState {
```

File: integtests/apitester/apitester.go [hunk 24]
```diff
@@ -42,158 +36,127 @@ func (c *Config) Run() error {
...
+		return nil
+	}
+	var respInst structs.ServiceInstance
+
+	err := json.Unmarshal(respData, &respInst)
+	if err != nil {
+		return errors.Wrap(err, "check responce unmarshal")
+	}
+	if respInst.LastOperation.State != expected.LastOperation.State {
+		return errors.New(fmt.Sprintf("wrong state. Have %s, want %s", respInst.LastOperation.State, expected.LastOperation.State))
```
Comment (dAdAbird) [inline integtests/apitester/apitester.go]:
```go
errors.Errof()
```

File: integtests/apitester/apitester.go [hunk 25]
```diff
@@ -42,158 +36,127 @@ func (c *Config) Run() error {
...
 	}
-	if respStatus != http.StatusOK || respStatus != http.StatusAccepted {
-		log.Println(respStatus)
-		return fmt.Errorf("Wrong status")
-	}
-	log.Println(respStatus, string(respData))
-	if len(testCase.ReqData) > 0 && string(respData) != testCase.RespData {
-		return fmt.Errorf("Wrong responce")
+	if respStatus != testCase.RespStatus {
+		return errors.New("wrong resp status. Have '" + strconv.Itoa(respStatus) + "', want " + strconv.Itoa(testCase.RespStatus))
```
Comment (dAdAbird) [inline integtests/apitester/apitester.go]: Make sense to use `errors.Errorf`

**Review by dAdAbird — CHANGES_REQUESTED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| pxc | failed |
| psmdb | failed |
| service-broker | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| pxc | failed |
| service-broker | passed |
| psmdb | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| psmdb | failed |
| service-broker | passed |
| pxc | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| service-broker | passed |
| psmdb | failed |
| pxc | failed |

**Review by dAdAbird — APPROVED** [review]

**Review by delgod — APPROVED** [review]

**Pull request merged**